src/lib/zf/tcp_out.c

```diff
@@ -1032,14 +1046,13 @@ static int tcp_output_segment(struct zf_tcp* tcp, struct tcp_seg *seg)
   /* Stop zwin timer */
   zf_tcp_timers_timer_stop(tcp, ZF_TCP_TIMER_ZWIN);
 
-  /* Do not retransmit a segment which has not got out of NIC. */
+  /* If we have not yet processed a TX completion for this segment, then the
+   * packet might still exist in a buffer, so we should send it either by PIO
+   * or by copying the packet to a new buffer. */
   if( seg->in_flight ) {
     zf_log_tcp_tx_trace(tcp, "%s: in-flight %u:%u\n", __func__,
                         tcp_seg_seq(seg), tcp_seg_seq(seg) + seg->len);
-    /* Make sure the segment is sent in fire and forget mode,
-     * that is either with PIO or with a buffer copy to be freed at
-     * tx complete */
-    send_flags = ZF_REQ_ID_PIO_FLAG | ZF_REQ_ID_PROTO_TCP_FREE | PKT_INVALID;
+    send_flags = ZF_REQ_ID_PIO_FLAG | ZF_REQ_ID_PROTO_TCP_KEEP | PKT_INVALID;
```

Comment:
The change here is the "important" one, to use `ZF_REQ_ID_PROTO_TCP_KEEP` instead of `ZF_REQ_ID_PROTO_TCP_FREE`. That way, [`zf_reactor_handle_tx_timestamp_event`](https://github.com/Xilinx-CNS/tcpdirect/blob/ba607b3ca2ada1e56f96042fdf8552012aa785c7/src/lib/zf/private/reactor.c#L82) will call `zf_tx_reports::complete`.